unic/ucd/normal/src/composition.rs

```diff
@@ -12,65 +12,35 @@
 
 use unic_utils::CharDataTable;
 
-
-pub struct Slice {
-    pub offset: u16,
-    pub length: u16,
-}
-
-fn bsearch_lookup_table<T>(
-    c: char,
-    r: &'static [(char, Slice)],
-    chars_table: &'static [T],
-) -> Option<&'static [T]> {
-    r.find(c).map(|slice| {
-        let offset = slice.offset as usize;
-        let length = slice.length as usize;
-        &chars_table[offset..(offset + length)]
-    })
-}
+use decomposition_type::DecompositionType;
 
 // == Canonical Composition (C) ==
-const CANONICAL_COMPOSITION_LOOKUP: &'static [(char, Slice)] =
-    include!("tables/canonical_composition_mapping_lookup.rsv");
-const CANONICAL_COMPOSITION_VALUES: &'static [(char, char)] =
-    include!("tables/canonical_composition_mapping_values.rsv");
+const CANONICAL_COMPOSITION_MAPPING: &'static [(char, &'static [(char, char)])] =
+    include!("tables/canonical_composition_mapping.rsv");
 
 /// Canonical Composition of the character.
 pub fn canonical_composition(c: char) -> Option<&'static ([(char, char)])> {
-    bsearch_lookup_table(
-        c,
-        CANONICAL_COMPOSITION_LOOKUP,
-        CANONICAL_COMPOSITION_VALUES,
-    )
+    CANONICAL_COMPOSITION_MAPPING.find(c).map(|it| *it)
 }
 
 // == Canonical Decomposition (D) ==
-const CANONICAL_DECOMPOSITION_LOOKUP: &'static [(char, Slice)] =
-    include!("tables/canonical_decomposition_mapping_lookup.rsv");
-const CANONICAL_DECOMPOSITION_VALUES: &'static [char] =
-    include!("tables/canonical_decomposition_mapping_values.rsv");
+const CANONICAL_DECOMPOSITION_MAPPING: &'static [(char, &'static [char])] =
+    include!("tables/canonical_decomposition_mapping.rsv");
 
 /// Canonical Decomposition of the character.
 pub fn canonical_decomposition(c: char) -> Option<&'static [char]> {
-    bsearch_lookup_table(
-        c,
-        CANONICAL_DECOMPOSITION_LOOKUP,
-        CANONICAL_DECOMPOSITION_VALUES,
-    )
+    CANONICAL_DECOMPOSITION_MAPPING.find(c).map(|it| *it)
 }
 
 // == Compatibility Decomposition (KD) ==
-const COMPATIBILITY_DECOMPOSITION_LOOKUP: &'static [(char, Slice)] =
-    include!("tables/compatibility_decomposition_mapping_lookup.rsv");
-const COMPATIBILITY_DECOMPOSITION_VALUES: &'static [char] =
-    include!("tables/compatibility_decomposition_mapping_values.rsv");
+use DecompositionType::*;
+use DecompositionType::Nobreak as NoBreak;
```

Comment:
nit: I would just fix this issue in the generation side, and don't special-case it here.

Looking at all the long_names for the value,
 looks safe to just *to_lower* and *capitalize* all DT value names. (http://www.unicode.org/Public/UCD/latest/ucd/PropertyValueAliases.txt)